pkg/validation/cluster.go

```diff
@@ -0,0 +1,129 @@
+package validation
+
+import (
+	"context"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/kubelet/config/v1beta1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/k8stopologyawareschedwg/deployer/pkg/validator"
+
+	mcov1 "github.com/openshift/machine-config-operator/pkg/apis/machineconfiguration.openshift.io/v1"
+
+	"github.com/openshift-kni/numaresources-operator/pkg/kubeletconfig"
+)
+
+func ValidateCluster(aContext context.Context, aClient client.Client) ([]validator.ValidationResult, error) {
+
+	var result []validator.ValidationResult
+
+	mcoKubeletConfigList := mcov1.KubeletConfigList{}
+	if err := aClient.List(aContext, &mcoKubeletConfigList); err != nil {
+		return nil, err
+	}
+
+	for _, mcoKubeletConfig := range mcoKubeletConfigList.Items {
+		kubeletConfig, err := kubeletconfig.MCOKubeletConfToKubeletConf(&mcoKubeletConfig)
+		if err != nil {
+			return nil, err
+		}
+
+		var nodeNames []string
+		machineConfigPools, err := getMCPsFromMCOKubeletConfig(aContext, aClient, mcoKubeletConfig)
+		if err != nil {
+			return nil, err
+		}
+		for _, mcp := range machineConfigPools {
+
+			nodes, err := getNodeListFromMachineConfigPool(aContext, aClient, *mcp)
```

Comment:
I'm not sure (inclined to nay) we need to check node by node.
`deployer` does in its own validation because it cannot depend on higher-order infra. It must work against plain kube, and in kube we only have nodes atm.
Here, OTOH, what we need to check is the `kubeletconfig` of a well-defined set of nodes. We (somehow) pick the set of nodes the `numaresourcesoperator` refers to (can be a union of sets), and we check that
1. it does exist a kubeletconfig referring to the same set
2. that kubeletconfig (its final rendering) has the setting we need

now, checking that the set of nodes the `numaresourcesoperator` refers to is the same set (or a subset of) the `kubeletconfig` refers to can be challenging. We may need to end up resolving the set to nodes and compare.

But the takeaway is we should still reason in terms of set of nodes, not individual nodes.

Reply:
So ... 
We should just care about the nodes with TAS enabled, that is those refered by the MCP in the NodeGroups of NRO, and those nodes should be referenced by a MCP of one MCO::kubeletConfig object with a "valid" (for TAS) kubelet configuration, 
right?